Python/Product/Analysis/LanguageServer/EditorFiles.cs

```diff
@@ -0,0 +1,257 @@
+                lock (_pendingChanges) {
+                    var notExpired = _pendingChanges
+                        .Where(p => p.textDocument.version.GetValueOrDefault() >= toVersion)
+                        .OrderBy(p => p.textDocument.version.GetValueOrDefault())
+                        .ToArray();
+
+                    _pendingChanges.Clear();
+                    if (notExpired.Any()) {
+                        next = notExpired.First();
+                        _pendingChanges.AddRange(notExpired.Skip(1));
+                    }
+                }
+                if (next.HasValue) {
+                    DidChangeTextDocument(next.Value, null);
+                }
+            } finally {
+                if (enqueueAction != null) {
+                    _server.TraceMessage($"Applied changes to {uri}");
+                    enqueueAction(doc);
+                }
+            }
+        }
+
+        public void UpdateParseDiagnostics(VersionCookie vc, Uri documentUri) {
+            lock (_lock) {
+                var last = _parseBufferDiagnostics;
+                Dictionary<Uri, PublishDiagnosticsEventArgs> newDiags = null;
+
+                foreach (var kv in vc.GetAllParts(documentUri)) {
+                    var part = _server.ProjectFiles.GetPart(kv.Key);
+                    if (!last.TryGetValue(part, out var lastVersion) || lastVersion.Version < kv.Value.Version || _ignoreDiagnosticsVersion) {
+                        last[part] = kv.Value;
+
+                        newDiags = newDiags ?? new Dictionary<Uri, PublishDiagnosticsEventArgs>();
+                        newDiags[kv.Key] = new PublishDiagnosticsEventArgs {
+                            uri = kv.Key,
+                            diagnostics = kv.Value.Diagnostics,
+                            _version = kv.Value.Version
+                        };
+                    }
+                }
+
+                if (newDiags != null) {
+                    _parseDiagnostics = newDiags;
+                }
+                _ignoreDiagnosticsVersion = false;
+
+                PublishDiagnostics(documentUri);
+            }
+        }
+
+        public void UpdateAnalysisDiagnostics(IProjectEntry projectEntry) {
+            var diags = _server.Analyzer.GetDiagnostics(projectEntry);
+            for (var i = 0; i < diags.Count; i++) {
+                diags[i].severity = _server.Settings.analysis.GetEffectiveSeverity(diags[i].code, diags[i].severity);
+            }
+
+            var severity = _server.Settings.analysis.GetEffectiveSeverity(ErrorMessages.UnresolvedImportCode, DiagnosticSeverity.Warning);
+            var pythonProjectEntry = projectEntry as IPythonProjectEntry;
+            var parse = pythonProjectEntry?.GetCurrentParse();
+
+            if (parse != null && _server.Settings.analysis.Show(severity)) {
+                var walker = new ImportStatementWalker(parse.Tree, pythonProjectEntry, _server.Analyzer, severity);
```

Comment:
If this walker is solely used to find missing imports, put that in its name.

If it's used during analysis, collect the diagnostics there instead of running it afterwards.